UI: move c-capnp to c++

deanlee opened the pull request:
move to capnproto

all function and logic is strictly the same as C version.

deanlee committed:
ui:move c-capnp to c++

deanlee committed:
rename reader to msg

include cereal/gen/cpp/log.capnp

deanlee committed:
move Subsockets into vector

deanlee committed:
fix some errors

restore some changes

restore previous

deanlee committed:
cleanup codes

cleanup codes

deanlee committed:
remove unused variable alert_size

deanlee commented:
Done, tested for a few miles.

geohot commented on selfdrive/ui/ui.cc:
```diff
@@ -387,207 +364,146 @@ static void update_status(UIState *s, int status) {
+    scene.networkType = data.getNetworkType();
+    scene.networkStrength = data.getNetworkStrength();
+    scene.batteryPercent = data.getBatteryPercent();
+    scene.batteryCharging = data.getBatteryStatus() == "Charging" ? true : false;
```
? true : false doesn't do anything

geohot commented on selfdrive/ui/ui.cc:
```diff
@@ -387,207 +364,146 @@ static void update_status(UIState *s, int status) {
+    scene.alert_text1 = data.getAlertText1();
+    scene.alert_text2 = data.getAlertText2();
+    scene.alert_ts = event.getLogMonoTime();
+    scene.alert_size = (uint8_t)data.getAlertSize();
```
Don't cast this, it's making an assumption of the numbers of the capnp enum. Instead, we should remove the ALERTSIZE_NONE... defines and use the cereal ones everywhere.

geohot commented on selfdrive/ui/sidebar.cc:
```diff
@@ -56,8 +55,8 @@ static void ui_draw_sidebar_network_type(UIState *s) {
 
-  if (s->scene.networkType <= 5) {
-    snprintf(network_type_str, sizeof(network_type_str), "%s", network_types[s->scene.networkType]);
+  if ((int)s->scene.networkType <= 5) {
```
Similar comment to below on the casts if there's a nice way to remove this one. We shouldn't make assumptions like "5"

deanlee replied:
Seriously， we should not use this enum value as index of array.

from the capnproto's document:

> Keep in mind when writing switch blocks that an enum read off the wire may have a numeric value that is not listed in its definition. This may be the case if the sender is using a newer version of the protocol, or if the message is corrupt or malicious.

quillford replied:
can we use a std::map?

deanlee replied:
sure we can. but I think switch is clear enough and is the "standard" way to do this type of thing.
use std::map can redule a few lines, but it  alloc memory from heap,and do a "hash table  search"  everytime to get the value from key.I really don't like that.

geohot replied:
Actually from a perf perspective I'm not sure std::map would be worse. switch statement = branch mispredict, idk if map would.

deanlee replied:
@quillford you'r right.std::map is much clearer than switch.

deanlee committed:
handle capnproto's enum in a robust way

add break to default

deanlee commented:
use switch to handle enum values now.

geohot commented on selfdrive/ui/paint.cc:
```diff
@@ -670,18 +663,33 @@ static void ui_draw_vision_footer(UIState *s) {
 
   const uint8_t *color = alert_colors[va_color];
-  const int alr_s = alert_sizes[va_size];
+  int alr_s = 0;
```
I find this much less readable. Does

const int alert_sizes[] = {		
   [cereal::ControlsState::AlertSize::NONE] = 0,		
   [cereal::ControlsState::AlertSize::SMALL] = 241,

work?

deanlee replied:
no, it didn't work in this case:
error: integral constant expression must have integral or unscoped enumeration type, not 'capnp::schemas::ThermalStatus_98c15adf4549151f'

deanlee committed:
switch -> std:map

deanlee committed:
use static std::map instead of switch

do cleanup

geohot commented on selfdrive/ui/sidebar.cc:
```diff
@@ -9,28 +10,32 @@ static void ui_draw_sidebar_background(UIState *s) {
 }
 
 static void ui_draw_sidebar_network_strength(UIState *s) {
+  static std::map<cereal::ThermalData::NetworkStrength, int> network_type_map = {
```
network_strength_map?

deanlee replied:
fixed -_-b

geohot commented:
Aside from that variable name, LGTM

deanlee committed:
fix wrong variable name

deanlee commented:
I checked again and there should be no problem

deanlee committed:
use FlatArrayMessageReader instead of custom MessageReader

remove messagehelp.h

Revert "use FlatArrayMessageReader instead of custom MessageReader"

This reverts commit 57d8b6b1e2b4bad908246f35eb068535b1627167.

use FlatArrayMessageReader instead of custom MessageReader

add header file

remove capnp_c lib,add kj lib

include serialize.h

fix

remove duplicate includes

deanlee commented on selfdrive/ui/ui.cc:
```diff
@@ -387,207 +363,150 @@ static void update_status(UIState *s, int status) {
+  auto amsg = kj::heapArray<capnp::word>((msg->getSize() / sizeof(capnp::word)) + 1);
+  memcpy(amsg.begin(), msg->getData(), msg->getSize());
+  capnp::FlatArrayMessageReader cmsg(amsg);
+  cereal::Event::Reader event = cmsg.getRoot<cereal::Event>();
```
use FlatArrayMessageReader instead of custom MessageReader

deanlee commented:
I'm done with this

geohot commented:
Sweet, LGTM
